source/frontend-packages.rst

```diff
@@ -0,0 +1,183 @@
+
+.. code-block:: javascript
+
+    import $ from 'jquery'
+    /* ... additional imports ... */
+
+    window.jQuery = window.$ = require('jquery')
+
+The first, native import will ensure jQuery is loaded early as a dependency. The second
+version uses the non-native ``require()`` function, which is the NodeJS format for
+module import before ES2015. This version is just a regular line of code and so it will
+*not* be hoisted and can properly inject the window globals ``jQuery`` and ``$``.
+
+Creating And Using Modules
+--------------------------
+
+New functionality may require you organize it into its own module. Use your best
+judgement here, but consider the same sort of reasoning we use to draw lines between
+functionality in Python. You can also group related modules into a package, by placing
+them in a directory.
+
+You can import your module to load it from the ``index.js``
+
+.. code-block:: javascript
+
+    import './utils.js'
+
+The ``./`` is necessary to distinguish from modules that would be loaded from the
+standard library or NPM installed packages. These imports are simply relative paths to
+the location of the module.
+
+You only need to include modules you might consider "top-level" in ``index.js``. For
+example, if you defined a helper function ``get_csrf_token()`` in that ``utils.js``
+and only needed to use that one function in another module, you would import it there,
+
+.. code-block:: javascript
+
+    import get_csrf_token from 'utils.js';
+
+This would assume that, in ``utils.js``, the function had been exported.
+
+.. code-block:: javascript
+
+    export function get_csrf_token() {
+        ...
+    }
+
+Installing New Packages
+=======================
+
+When adding new frontend dependencies you should find an NPM packaged distribution
+of the version you need. Preferably the vendor or project will manage this, but the
+Javascript world is still catching up to proper packaging and you may find third-party
+distributions which are also acceptable.
+
+You can install an NPM package both locally and configured in your package.json
+with a single command::
+
+    npm install --save package-name==1.2.3
+
+Upgrading Existing Packages
+===========================
+
+Periodically we need to look at upgrading the versions of third-party packages we
+depend on. This includes multiple steps.
+
+* Identify new versions of packages we use.
+* Upgrade to a newer version of a package and successfully test that no regressions occur.
+* Update ``package.json`` with the new version and commit this change.
+
+We can make this process a little easier, and we can also enforce some rules about them
+thanks to NPMs. When specifying package versions in ``package.json`` we can tell NPM about
+how we want it to interpret the version number we give.
+
+* ``1.2`` installs the newest ``1.2`` patch release
+* ``1.2.x`` or ``~1.2.3`` do the same, but the later defines the original minimum specified
+  when the package version was last pinned.
```

Comment:
It sounds like `1.2` and `1.2.x` mean the same thing, so maybe `1.2.x` should be included in the first line instead of the second.

Also, I don't quite follow the explanation of `~1.2.3`.  Is it a purely cosmetic difference to provide documentation that when you wrote this, the current version was `1.2.3`?